src/constraints.jl

```diff
@@ -0,0 +1,559 @@
+        model::JuMP.Model,
+        con::_DisjunctConstraint,
+        name::String = ""
+    )::DisjunctConstraintRef
+
+Extend `JuMP.add_constraint` to add a [`_DisjunctConstraint`](@ref) to a [`GDPModel`](@ref). 
+The constraint is added to the `GDPData` in the `.ext` dictionary of the `GDPModel`.
+"""
+function JuMP.add_constraint(
+    model::JuMP.Model,
+    con::_DisjunctConstraint,
+    name::String = ""
+)
+    is_gdp_model(model) || error("Can only add disjunct constraints to `GDPModel`s.")
+    data = ConstraintData(con.constr, name)
+    idx = _MOIUC.add_item(gdp_data(model).disjunct_constraints, data)
+    _add_indicator_var(con, idx, model)
+    return DisjunctConstraintRef(model, idx)
+end
+
+################################################################################
+#                              DISJUNCTIONS
+################################################################################
+function _process_structure(_error, s::Vector{LogicalVariableRef}, model::JuMP.Model)
+    allunique(s) ||_error("Not all the logical indicator variables are unique.")
+    for lvref in s
+        if !JuMP.is_valid(model, lvref)
+            _error("`$lvref` is not a valid logical variable reference.")
+        elseif !haskey(_indicator_to_constraints(model), JuMP.index(lvref))
+            _error("`$lvref` is not associated with any constraints.")
+        end
+    end
+    return s
+end
+
+# fallback
+function _process_structure(_error, s, model::JuMP.Model)
+    _error("Unrecognized disjunction input structure.") # TODO add details on proper syntax
+end
+
+# Write the main function for creating disjunctions that is macro friendly
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure::Vector,
+    name::String
+)
+    is_gdp_model(model) || error("Can only add disjunctions to `GDPModel`s.")
+
+    # build the disjunction
+    indicators = _process_structure(_error, structure, model)
+    disjunction = Disjunction(JuMP.index.(indicators))
+
+    # add it to the model
+    disjunction_data = ConstraintData(disjunction, name)
+    idx = _MOIUC.add_item(_disjunctions(model), disjunction_data)
+
+    _set_ready_to_optimize(model, false)
+    return DisjunctionRef(model, idx)
+end
+
+# Fallback disjunction build for nonvector structure
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure,
+    name::String
+)
+    _error("Unrecognized disjunction input structure.")
+end
+
+# Disjunction build for nested disjunctions with indicator variable given
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure,
+    name::String,
+    tag::DisjunctConstraint
+)
+    dref = _disjunction(_error, model, structure, name)
+    obj = JuMP.constraint_object(dref)
+    return JuMP.add_constraint(model, _DisjunctConstraint(obj, tag.indicator), name)
```

Comment:
Are we ok with storing nested disjunctions twice?

Reply:
I haven't looked into this code on how nesting will work. I'll play around a bit with it. However, we might want to leave nesting for a separate PR. What are your thoughts?

Reply:
My thoughts are that it would be best to go ahead and support nesting since doing so might require breaking changes. The current definition structure should work. The only caveat is we store it twice currently and I am not sure if the reformulations are correct with it.

Reply:
I am guessing we need a way of knowing if a disjunction is nested or not for reformulating

Reply:
Yes, we need to be able to reformulate nested disjunctions first so that we can add the reformulation as a disjunct constraint(s) and then reformulate the upper level disjunctions...

Reply:
It get's tricky once we have multiple levels because we have to work our way from the inner most disjunction. Almost sounds like we should have some object tree tracking the nested relationship (dependency graph)...

Reply:
Ya this is going to need some careful thought. I would be hesitant to add to disjunct constraints while reformulating since that will make it tricky to change to anothet reformulation. We can maybe make some buffer field to store them as we build the reformulation. Also note that th e disjunctions are stored in creation order and the inner most must be created first so they will be reformulated first.

Reply:
After thinking about it more, I think we can do the reformulations from the top down via recursion which will in turn evaluate it from the inside out. We should just store whether a disjunction is nested or not. If it isn't then reformulate it and it's underlying disjunctions otherwise skip it. We can set up a call to discuss this further if you want

Reply:
done